### PR TITLE
Conversion to PythonCall

### PR DESCRIPTION
**Pull request by tylerjthomas9**
Here is my initial attempt at dropping ScikitLearn.jl, and just including the base PythonCall.jl wrapper. Let me know what you think. I think that it would be great to methods for preparing the data, similar to what we did in CatBoost.jl, but I haven't implemented them yet.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -44,15 +43,8 @@ jobs:
...
       - uses: julia-actions/julia-buildpkg@v1
-# The following is needed for Julia <=0.8.4 on Linux OS 
-# due to old version of libstcxx used by Julia
-      - name: "Export LD_LIBRARY_PATH envrioment variable"
-        if: ${{matrix.version == '1.6'}}
-        run: echo "LD_LIBRARY_PATH=/home/runner/.julia/conda/3/x86_64/lib" >> $GITHUB_ENV
       - uses: julia-actions/julia-runtest@v1
-        env:
-          LD_LIBRARY_PATH: /home/runner/.julia/conda/3/x86_64/lib
       - uses: julia-actions/julia-processcoverage@v1
```
Comment (ablaom) [inline .github/workflows/ci.yml:59]: Are we sure this is not still needed? It would seem to me that this issue is orthogonal to the PythonCall/PyCall choice. 

@OkonSamuel What say you?

File: src/ScikitLearnAPI.jl [hunk 2]
```diff
@@ -0,0 +1,52 @@
+
+module ScikitLearnAPI
+
+
+using Tables
```
Comment (ablaom) [inline src/ScikitLearnAPI.jl]: Curious, but I couldn't see where Tables is used in this module.

File: src/ScikitLearnAPI.jl [hunk 3]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (ablaom) [inline src/ScikitLearnAPI.jl]: Perhaps we could add some kind of acknowledgement here, assuming you agree it's accurate:
```suggestion
# The mapping of Julia functions into python functions in this file is adapted from code in ScikitLearn.jl, written by Cédric St-Jean. 
```

File: test/generic_api_tests.jl [hunk 4]
```diff
@@ -12,7 +12,7 @@ import MLJBase: finaltypes,
     Table
 
 const MODELS = filter(finaltypes(Model)) do m
-    package_name(m) == "ScikitLearn"
+    package_name(m) == "MLJScikitLearnInterface"
 end
```
Comment (ablaom) [inline test/generic_api_tests.jl:16]: Yes!

File: test/models/clustering.jl [hunk 5]
```diff
@@ -17,22 +17,23 @@ fparams = (
 models = (
         AffinityPropagation,
         AgglomerativeClustering,
+        Birch, 
         DBSCAN,
```
Comment (ablaom) [inline test/models/clustering.jl:22]: That's a good catch. @OkonSamuel Can you add this already to your  PR, #55?

File: src/models/clustering.jl [hunk 6]
```diff
@@ -30,22 +30,22 @@ meta(AffinityPropagation,
 const AgglomerativeClustering_ = skcl(:AgglomerativeClustering)
 @sk_uns mutable struct AgglomerativeClustering <: MMI.Unsupervised
     n_clusters::Int     = 2::(_ ≥ 1)
-    affinity::String    = "euclidean"::(_ in ("euclidean", "l1", "l2", "manhattan", "cosine", "precomputed"))
+    metric::String      = "euclidean"::(_ in ("euclidean", "l1", "l2", "manhattan", "cosine", "precomputed"))
```
Comment (ablaom) [inline src/models/clustering.jl]: @OkonSamuel It seems this change is not in your PR. Can you comment?

File: src/models/clustering.jl [hunk 7]
```diff
@@ -143,23 +143,23 @@ const FeatureAgglomeration_ = skcl(:FeatureAgglomeration)
     connectivity::Any      = nothing
     # XXX unclear how to pass a proper callable here; just passing mean = nok
     # pooling_func::Function = mean
-    affinity::Any          = "euclidean"::(_ isa Function || _ in ("euclidean", "l1", "l2", "manhattan", "cosine",  "precomputed"))
+    metric::Any            = "euclidean"::(_ isa Function || _ in ("euclidean", "l1", "l2", "manhattan", "cosine",  "precomputed"))
     compute_full_tree::Union{String,Bool} = "auto"::(_ isa Bool || _ == "auto")
```
Comment (ablaom) [inline src/models/clustering.jl]: @OkonSamuel

**Review by ablaom — COMMENTED** [review]
Fantastic work. I can't believe how quickly you pulled this together.

I note that CI is failing on 1.6 Ubuntu. My guess is that a different version of the python libraries are being installed, possible related to @OkonSamuel's libstcxx hack that has been removed??

Is there a way to restrict sk-learn version to [1.2, 1.3)?

File: src/models/clustering.jl (hunk 6 above)
Comment (OkonSamuel) [reply src/models/clustering.jl]: It's not. I'm aware of this but choose to leave the `affinity` parameter till the next minor release of python sklearn.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -23,8 +20,10 @@ jobs:
         version:
           - '1.6'
           - '1'
+          - 'nightly'
         os:
           - ubuntu-latest
+          - windows-latest
```
Comment (OkonSamuel) [inline .github/workflows/ci.yml:27]: I think we also need to add a test for MacOs also

**Comment by tylerjthomas9** [comment]
> Fantastic work. I can't believe how quickly you pulled this together.
> 
> I note that CI is failing on 1.6 Ubuntu. My guess is that a different version of the python libraries are being installed, possible related to @OkonSamuel's libstcxx hack that has been removed??
> 
> Is there a way to restrict sk-learn version to [1.2, 1.3)?

On Julia 1.6, scikit-learn v1.1.1 is being installed due to `libstdcxx` restrictions. I can limit the `CondaPkg.toml` bounds for `scikit-learn`, but this would prevent `v1.6` from being able to use the package. I will investigate and see if we can just patch the scikit-learn v1.1.1 issue.

File: src/models/clustering.jl (hunk 6 above)
Comment (ablaom) [reply src/models/clustering.jl:33]: @OkonSamuel suggests leaving this as `affinity` and living with the dep warnings until sk-learn 1.4 is released.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (tylerjthomas9) [reply .github/workflows/ci.yml:59]: I think that this hack was due to outdated `libstdcxx` libraries in julia. `PythonCall.jl` handles this now, and restricts python package versions to only those that are compatible with the current julia session. 

edit: this may be needed still for scikit-learn v1.2 to install on older Julia versions. Trying to figure it out.

File: src/ScikitLearnAPI.jl (hunk 2 above)
Comment (tylerjthomas9) [reply src/ScikitLearnAPI.jl]: Removed. I forgot to remove it when converting the ScikitLearn.jl code. Good catch.

File: src/models/clustering.jl (hunk 6 above)
Comment (tylerjthomas9) [reply src/models/clustering.jl]: I reverted these changes. It is probably worth doing a different PR down the road to make these changes.

**Comment by ablaom** [comment]
If you haven't done so already, perhaps you should rebase off the new dev, now that @OkonSamuel's PR is merged?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (ablaom) [reply .github/workflows/ci.yml:27]: Agreed!

**Comment by OkonSamuel** [comment]
> > Fantastic work. I can't believe how quickly you pulled this together.
> > I note that CI is failing on 1.6 Ubuntu. My guess is that a different version of the python libraries are being installed, possible related to @OkonSamuel's libstcxx hack that has been removed??
> > Is there a way to restrict sk-learn version to [1.2, 1.3)?
> 
> On Julia 1.6, scikit-learn v1.1.1 is being installed due to `libstdcxx` restrictions. I can limit the `CondaPkg.toml` bounds for `scikit-learn`, but this would prevent `v1.6` from being able to use the package. I will investigate and see if we can just patch the scikit-learn v1.1.1 issue.

That's what my hack was meant to address for the CI.
We could just add a readme similar to, https://github.com/cstjean/ScikitLearn.jl#known-issue

**Comment by tylerjthomas9** [comment]
> If you haven't done so already, perhaps you should rebase off the new dev, now that @OkonSamuel's PR is merged?

I think that PR has been merged into my branch. I see the changes he made. I will have time after Tuesday of this week to do any remaining tasks to get this PR ready.

**Comment by tylerjthomas9** [comment]
I added a hack to get the CI working on Ubuntu + Julia 1.6. The hack just replaces julia's libstdcxx with the action runner's version, which is more up to date. I am not sure what is different about the github actions version of julia, but it has a older libstdcxx version than julia 1.6 when I install it on my personal computer. For me, the test suite passes with 0 issues on local julia 1.6 installs. 

Let me know what you think.

**Comment by ablaom** [comment]
@tylerjthomas9 I can't see anything left to do here but I'd like @OkonSamuel to comment on your alternative hack. 

Then we should coordinate this merge with the OutlierDetectionPython.jl PR. I'll open a tracking issue shortly.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -23,36 +20,43 @@ jobs:
...
+        # due to old version of libstcxx used by Julia
+        # taken from https://github.com/hhaensel/ReplaceLibstdcxx.jl/blob/main/src/ReplaceLibstdcxx.jl
+        run: |
+              julia -e '
+                libs = filter(x -> ! occursin("32", x), getindex.(split.(readlines(pipeline(`ldconfig -p`, `grep libstdc`)), r"\s*=>\s*"), 2))
+                source_dir = dirname(libs[end])
+                julia_lib_dir = joinpath(dirname(Sys.BINDIR), "lib", "julia")
+                julia_lib_file = get(filter(endswith(r"libstdc\+\+.so\.\d+\.\d+\.\d+"), readdir(julia_lib_dir, join = true)), 1, nothing)
+                julia_lib_version = match(r"so(\.\d+)\.", julia_lib_file).captures[1]
+                source_lib = get(filter(endswith(r"libstdc\+\+.so\.\d+\.\d+\.\d+"), readdir(source_dir, join = true)), 1, nothing)
```
Comment (OkonSamuel) [inline .github/workflows/ci.yml:49]: All these is dependent on  `libs` not being empty. Which I think makes sense for Ubuntu builds since there is always a `libstdc` shared library. Right?

**Comment by OkonSamuel** [comment]
> @tylerjthomas9 I can't see anything left to do here but I'd like @OkonSamuel to comment on your alternative hack.
> 
> Then we should coordinate this merge with the OutlierDetectionPython.jl PR. I'll open a tracking issue shortly.

It's the same thing I did. i.e replacing Julia's libstdxx with the up to date system's libstdxx (In my case I replaced it with the libstdxx from the Conda package). Can we document this Hack in the README?

> I added a hack to get the CI working on Ubuntu + Julia 1.6. The hack just replaces julia's libstdcxx with the action runner's version, which is more up to date. I am not sure what is different about the github actions version of julia, but it has a older libstdcxx version than julia 1.6 when I install it on my personal computer. For me, the test suite passes with 0 issues on local julia 1.6 installs.
> 
> Let me know what you think.

Are you running Ubuntu?

**Comment by tylerjthomas9** [comment]
Added a quick note in the documentation for fixing the `libstdcxx` issues on Linux. Let me know if I should change anything. 

> Are you running Ubuntu?

I am. I also tried github codespaces, and had 0 issues with Julia 1.6.7 on Linux.

**Review by OkonSamuel — APPROVED** [review]
LGTM!!!

**Comment by OkonSamuel** [comment]
> Added a quick note in the documentation for fixing the `libstdcxx` issues on Linux. Let me know if I should change anything.
> 
> > Are you running Ubuntu?
> 
> I am. I also tried github codespaces, and had 0 issues with Julia 1.6.7 on Linux.

That's strange. It fails when I try it on my ubuntu system. Although, I haven't ran any update on my ubuntu for a while now.
